### PR TITLE
Add Support for UserDefinedType in RestAPIv2

### PR DESCRIPTION
**Pull request by clun**
This pull request introduces support for `UserDefinedType` in Api REST v2 (schemas)

**Jenv Support**
As of now the project required a JDK8 explicitly (or we got standard compilation issues about class versionning 53.0 vs 55.0). I use `jenv` to handle multiple JDK on my laptop. To enforce JDK8 jenv create a `.java-version` file at root. We do not want it in the githup repo so I amended the `.gitignore` file.

**Support for compilation in Eclipse IDE**
- Multiple plugin executions are not recignize by m2e in eclipse and turn the project in error. I introduced a new block in `PluginManagement` to exclude those using `org.eclipse.m2e.lifecycle-mapping` no side effect for other languages.

- For the compiler plugin I add to specialized the `org.immutables` annotation processor to avoid conflicts

**UserDefineTypes Support**

- New class `UserDefinedTypesResource` exposing the endpoint `@Path("/v2/schemas/keyspaces/{keyspaceName}/types")`

- Providing findAll(), findById(), create(), delete() but no update (not available for a UDT)

- Creating new beans like `UserDefinedTypeAdd` and `ColumnDefinitionUserDefinedType` accordingly

- Mapping of the class in swagger UI in `io.stargate.web.impl.Server`

**Test Dataset**

To test the endpoint i used the following objects:

```sql
CREATE TYPE IF NOT EXISTS udt0 (
  width   int,
  height  int
);

CREATE TYPE IF NOT EXISTS udt1 (
  field_uid  uuid,
  field_text text,
  field_int  int,
  field_ts   timestamp,
  field_date timeuuid,
  field_set  set <text>,
  field_list list<int>,
  field_map  map<text,text>,
  field_udt1 frozen<udt0>,
  field_udt2 map<text,frozen<udt0>>
);
```
 
**Test Methodology**
- On Mac make sure to create the alias
```bash
sudo ifconfig lo0 alias 127.0.0.2
```
- Create a ccm cluster
```bash
rm -Rf ~/.ccm/stargate
ccm create stargate -v 3.11.6 -n 1 -s -b
```
- Start stargate

```bash
./starctl --cluster-name stargate --cluster-seed 127.0.0.1 --cluster-version 3.11 --listen 127.0.0.2 --bind-to-listen-address --simple-snitch
```
- Get an authentication token
```bash
curl -X POST "http://127.0.0.2:8081/v1/auth" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{  \"username\": \"cassandra\",  \"password\": \"cassandra\"}"
```

```json
{
  "authToken": "a451d876-f56d-4190-a939-0fe2f7f909d9"
}
```

- Create a keyspace `ks_udt`

```bash
curl -X POST "http://127.0.0.2:8082/v2/schemas/keyspaces" -H  "accept: application/json" -H  "X-Cassandra-Token: a451d876-f56d-4190-a939-0fe2f7f909d9" -H  "Content-Type: application/json" -d "{ \"name\": \"ks_udt\", \"replicas\": 1}"
```

```json
{
  "name": "ks_udt"
}
```

- Create the User Defined type `udt0`

*JSON Payload:*
```json
{
    "name": "udt0",
    "ifNotExists": true,
    "columnDefinitions": [
      {
        "name": "width",
        "typeDefinition": "int"
      },
      {
        "name": "height",
        "typeDefinition": "int"
      }
    ]
}
```

*Curl command*
```
curl -X POST "http://127.0.0.2:8082/v2/schemas/keyspaces/ks_udt/types" -H  "accept: application/json" -H  "X-Cassandra-Token: a451d876-f56d-4190-a939-0fe2f7f909d9" -H  "Content-Type: application/json" -d "{    \"name\": \"udt0\",    \"ifNotExists\": true,    \"columnDefinitions\": [      {        \"name\": \"width\",        \"typeDefinition\": \"int\"      },      {        \"name\": \"height\",        \"typeDefinition\": \"int\"      }    ]}"
```

*Output*
```
{
  "name": "udt0"
}
```

- Create the User Defined type `udt1`

*JSON Payload:*
```json
{
    "name": "udt1",
    "ifNotExists": true,
    "columnDefinitions": [
      {
        "name": "field_uid",
        "typeDefinition": "uuid"
      },
      {
        "name": "field_text",
        "typeDefinition": "varchar"
      },
      {
        "name": "field_int",
        "typeDefinition": "int"
      },
      {
        "name": "field_ts",
        "typeDefinition": "timestamp"
      },
      {
        "name": "field_date",
        "typeDefinition": "timeuuid"
      },
      {
        "name": "field_set",
        "typeDefinition": "set<varchar>"
      },
      {
        "name": "field_list",
        "typeDefinition": "list<int>"
      },
      {
        "name": "field_map",
        "typeDefinition": "map<varchar, varchar>"
      },
      {
        "name": "field_udt1",
        "typeDefinition": "frozen<udt0>"
      },
      {
        "name": "field_udt2",
        "typeDefinition": "map<varchar, frozen<udt0>>"
      }
    ]
  }
```

*Curl command*
```
curl -X POST "http://127.0.0.2:8082/v2/schemas/keyspaces/ks_udt/types" -H  "accept: application/json" -H  "X-Cassandra-Token: a451d876-f56d-4190-a939-0fe2f7f909d9" -H  "Content-Type: application/json" -d "{    \"name\": \"udt1\",    \"ifNotExists\": true,    \"columnDefinitions\": [      {        \"name\": \"field_uid\",        \"typeDefinition\": \"uuid\"      },      {        \"name\": \"field_text\",        \"typeDefinition\": \"varchar\"      },      {        \"name\": \"field_int\",        \"typeDefinition\": \"int\"      },      {        \"name\": \"field_ts\",        \"typeDefinition\": \"timestamp\"      },      {        \"name\": \"field_date\",        \"typeDefinition\": \"timeuuid\"      },      {        \"name\": \"field_set\",        \"typeDefinition\": \"set<varchar>\"      },      {        \"name\": \"field_list\",        \"typeDefinition\": \"list<int>\"      },      {        \"name\": \"field_map\",        \"typeDefinition\": \"map<varchar, varchar>\"      },      {        \"name\": \"field_udt1\",        \"typeDefinition\": \"frozen<udt0>\"      },      {        \"name\": \"field_udt2\",        \"typeDefinition\": \"map<varchar, frozen<udt0>>\"      }    ]  }"
```

*Output*
```
{
  "name": "udt1"
}
```

- List all UDTs for a keyspace

*Command*
```
curl -X GET "http://127.0.0.2:8082/v2/schemas/keyspaces/ks_udt/types" -H  "accept: application/json" -H  "X-Cassandra-Token: 69239c7d-db06-45e2-bbe8-719877112c2c"
```

*Expected Output*
```
{
  "data": [
    {
      "name": "udt0",
      "keyspace": "ks_udt",
      "columnDefinitions": [
        {
          "name": "width",
          "typeDefinition": "int"
        },
        {
          "name": "height",
          "typeDefinition": "int"
        }
      ]
    },
    {
      "name": "udt1",
      "keyspace": "ks_udt",
      "columnDefinitions": [
        {
          "name": "field_uid",
          "typeDefinition": "uuid"
        },
        {
          "name": "field_text",
          "typeDefinition": "varchar"
        },
        {
          "name": "field_int",
          "typeDefinition": "int"
        },
        {
          "name": "field_ts",
          "typeDefinition": "timestamp"
        },
        {
          "name": "field_date",
          "typeDefinition": "timeuuid"
        },
        {
          "name": "field_set",
          "typeDefinition": "set<varchar>"
        },
        {
          "name": "field_list",
          "typeDefinition": "list<int>"
        },
        {
          "name": "field_map",
          "typeDefinition": "map<varchar, varchar>"
        },
        {
          "name": "field_udt1",
          "typeDefinition": "frozen<udt0>"
        },
        {
          "name": "field_udt2",
          "typeDefinition": "map<varchar, frozen<udt0>>"
        }
      ]
    }
  ]
}
```

- Find by id (OK=200)

*Command*
```
curl -X GET "http://127.0.0.2:8082/v2/schemas/keyspaces/ks_udt/types/udt0" -H  "accept: application/json" -H  "X-Cassandra-Token: 69239c7d-db06-45e2-bbe8-719877112c2c"
```

*Expected Output*
```
{
  "data": {
    "name": "udt0",
    "keyspace": "ks_udt",
    "columnDefinitions": [
      {
        "name": "width",
        "typeDefinition": "int"
      },
      {
        "name": "height",
        "typeDefinition": "int"
      }
    ]
  }
}
```

- Find by id with non valid udt name should return 404

*Command*
```
curl -X GET "http://127.0.0.2:8082/v2/schemas/keyspaces/ks_udt/types/I_dont_exist" -H  "accept: application/json" -H  "X-Cassandra-Token: 69239c7d-db06-45e2-bbe8-719877112c2c"
```

*output*
```
{
  "description": "Resource not found: table 'I_dont_exist' not found",
  "code": 404
}
```

- Delete by id with KO parameter, I cannot delete idt0 as it is used by udt1 expected 400

*Command*
```
curl -X DELETE "http://127.0.0.2:8082/v2/schemas/keyspaces/ks_udt/types/udt0" -H  "accept: application/json" -H  "X-Cassandra-Token: 69239c7d-db06-45e2-bbe8-719877112c2c"
```

*output*
```
{
  "description": "Bad request: org.apache.cassandra.stargate.exceptions.InvalidRequestException: Cannot drop user type ks_udt.udt0 as it is still used by user type udt1",
  "code": 400
}
```

- Delete by id OK
*command*
```
curl -X DELETE "http://127.0.0.2:8082/v2/schemas/keyspaces/ks_udt/types/udt1" -H  "accept: application/json" -H  "X-Cassandra-Token: 69239c7d-db06-45e2-bbe8-719877112c2c"
```

*output*
No Payload and code 204.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -413,6 +461,53 @@
...
+	           <pluginExecution>
+			    <pluginExecutionFilter>
+				 <groupId>au.com.acegi</groupId>
+				 <artifactId>xml-format-maven-plugin</artifactId>
+				 <versionRange>[3.1,)</versionRange>
+				 <goals><goal>xml-check</goal></goals>
+			   </pluginExecutionFilter>
+			   <action><execute /></action>
+			 </pluginExecution>
+	         <pluginExecution>
```
Comment (dougwettlaufer) [inline pom.xml:481]: There's a good chance CI is going to fail on this due to formatting. From command line you'll want to run `mvn xml-format:xml-format fmt:format` and repush.

**Comment by dougwettlaufer** [comment]
Build failure

```java
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ persistence-api ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 131 source files to /workspace/persistence-api/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /workspace/persistence-api/src/main/java/io/stargate/db/schema/Schema.java:[54,17] cannot find symbol
  symbol:   class SchemaBuilder
  location: class io.stargate.db.schema.Schema
[ERROR] /workspace/persistence-api/src/main/java/io/stargate/db/schema/Schema.java:[58,17] cannot find symbol
  symbol:   class SchemaBuilder
  location: class io.stargate.db.schema.Schema
[ERROR] /workspace/persistence-api/src/main/java/io/stargate/db/datastore/DataStore.java:[23,36] cannot find symbol
  symbol:   class QueryBuilder
  location: package io.stargate.db.query.builder
[ERROR] /workspace/persistence-api/src/main/java/io/stargate/db/datastore/DataStore.java:[39,11] cannot find symbol
  symbol:   class QueryBuilder
  location: interface io.stargate.db.datastore.DataStore
[INFO] 4 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for io.stargate:stargate 1.0.17-SNAPSHOT:
[INFO] 
[INFO] io.stargate:stargate ............................... SUCCESS [  0.801 s]
[INFO] core ............................................... SUCCESS [ 13.088 s]
[INFO] stargate-starter ................................... SUCCESS [  6.906 s]
[INFO] persistence-api .................................... FAILURE [  2.777 s]
[INFO] persistence-common ................................. SKIPPED
[INFO] persistence-test ................................... SKIPPED
[INFO] authnz ............................................. SKIPPED
[INFO] persistence-cassandra-3.11 ......................... SKIPPED
[INFO] persistence-cassandra-4.0 .......................... SKIPPED
[INFO] persistence-dse-6.8 ................................ SKIPPED
[INFO] cql ................................................ SKIPPED
[INFO] graphqlapi ......................................... SKIPPED
[INFO] restapi ............................................ SKIPPED
[INFO] auth-api ........................................... SKIPPED
[INFO] auth-table-based-service ........................... SKIPPED
[INFO] auth-jwt-service ................................... SKIPPED
[INFO] health-checker ..................................... SKIPPED
[INFO] config-store-api ................................... SKIPPED
[INFO] config-store-yaml .................................. SKIPPED
[INFO] rate-limiting-global ............................... SKIPPED
[INFO] testing-services ................................... SKIPPED
[INFO] testing ............................................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  25.171 s
[INFO] Finished at: 2021-03-25T19:33:42Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project persistence-api: Compilation failure: Compilation failure: 
[ERROR] /workspace/persistence-api/src/main/java/io/stargate/db/schema/Schema.java:[54,17] cannot find symbol
[ERROR]   symbol:   class SchemaBuilder
[ERROR]   location: class io.stargate.db.schema.Schema
[ERROR] /workspace/persistence-api/src/main/java/io/stargate/db/schema/Schema.java:[58,17] cannot find symbol
[ERROR]   symbol:   class SchemaBuilder
[ERROR]   location: class io.stargate.db.schema.Schema
[ERROR] /workspace/persistence-api/src/main/java/io/stargate/db/datastore/DataStore.java:[23,36] cannot find symbol
[ERROR]   symbol:   class QueryBuilder
[ERROR]   location: package io.stargate.db.query.builder
[ERROR] /workspace/persistence-api/src/main/java/io/stargate/db/datastore/DataStore.java:[39,11] cannot find symbol
[ERROR]   symbol:   class QueryBuilder
[ERROR]   location: interface io.stargate.db.datastore.DataStore
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

**Review by clun — COMMENTED** [review]
Changing name and adding the support for UserDefined Type
Screen Shot 2021-03-25 at 19 10 12

**Comment by clun** [comment]
First comment has been updated with test procedure and use case, PR has been clean up

File: pom.xml [hunk 2]
```diff
@@ -355,6 +356,15 @@
...
             <failOnWarning>false</failOnWarning>
+            <!-- 
+            <annotationProcessorPaths>
+             <path>
+              <groupId>org.immutables</groupId>
+              <artifactId>value</artifactId>
+              <version>${immutables.version}</version>
+              </path>
+            </annotationProcessorPaths>
+            -->
```
Comment (dougwettlaufer) [inline pom.xml:367]: Can this just be removed?

Comment (clun) [reply pom.xml:367]: Eggs and chicken:
- needs to be uncomment to work in eclipse. 
- need to be comment for the generator to run and create the builder. 

I would like to keep it because i do not found another way to have it running on the IDE.

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+            return Response.status(Response.Status.BAD_REQUEST)
+                .entity(
+                    new Error(
+                        "Type name must be provided", Response.Status.BAD_REQUEST.getStatusCode()))
+                .build();
+          }
+          List<Column> columns = new ArrayList<>();
+          for (ColumnDefinitionUserDefinedType colDef : udtAdd.getColumnDefinitions()) {
+            String columnName = colDef.getName();
+            if (columnName == null || columnName.equals("")) {
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:270]:
```suggestion
            if (Strings.isNullOrEmpty(columnName)) {
```

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+          db.getAuthorizationService()
+              .authorizeSchemaWrite(
+                  authenticatedDB.getAuthenticationSubject(),
+                  keyspaceName,
+                  null,
+                  Scope.CREATE,
+                  SourceAPI.REST,
+                  ResourceKind.TYPE);
+          String typeName = udtAdd.getName();
+          if (typeName == null || typeName.equals("")) {
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:260]:
```suggestion
          if (Strings.isNullOrEmpty(columnName)) {
```

**Comment by eribeiro** [comment]
The PR is lacking integration tests.

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+
+          // Permissions on a type are the same as keyspace
+          // (CreateTypeFetcher for graphQl API)
+          db.getAuthorizationService()
+              .authorizeSchemaRead(
+                  authenticatedDB.getAuthenticationSubject(),
+                  Collections.singletonList(keyspaceName),
+                  null,
+                  SourceAPI.REST,
+                  ResourceKind.TYPE);
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:347]: As we are deleting a type, I think a more sensible approach would be to check for schema write permissions as below:
```
          db.getAuthorizationService()
              .authorizeSchemaWrite(
                  authenticatedDB.getAuthenticationSubject(),
                  keyspaceName,
                  null,
                  Scope.DROP,
                  SourceAPI.REST,
                  ResourceKind.TYPE);
```

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+                  .map(this::mapUdtAsReponse)
+                  .collect(Collectors.toList());
+          // Filter based on user credentials
+          db.getAuthorizationService()
+              .authorizeSchemaRead(
+                  authenticatedDB.getAuthenticationSubject(),
+                  Collections.singletonList(keyspaceName),
+                  udtResponses.stream()
+                      .map(UserDefinedTypeResponse::getName)
+                      .collect(Collectors.toList()),
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:132]: Sincere question: wouldn't suffice to check if the user has schema read privilege to the keyspace? There's really a need of passing this list here or could we just pass `null`?

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+          Keyspace keyspace = authenticatedDB.getDataStore().schema().keyspace(keyspaceName);
+
+          // Parameters validations, Keyspacename, type
+          if (keyspace == null) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                .entity(
+                    new Error(
+                        "keyspace does not exists", Response.Status.BAD_REQUEST.getStatusCode()))
+                .build();
+          }
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:250]: You are checking if keyspace exists here, but doesn't do the same in `delete` method. Wouldn't be the case of using the same logic there?

**Comment by eribeiro** [comment]
The lacking of integration tests is the strongest thing holding it back, I guess. Maybe some unit tests for the portions where it is suitable wouldn't hurt too.

File: .gitignore [hunk 8]
```diff
@@ -47,3 +47,5 @@ settings.xml
 .classpath
 *.prefs
 .factorypath
+# JENV
+.java-version
```
Comment (eribeiro) [inline .gitignore:51]: This is unrelated to the PR. Shouldn't it be part of other PR/commit?

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+    return RequestHandler.handle(
+        () -> {
+          // Accessing datastore
+          AuthenticatedDB authenticatedDB = db.getDataStoreForToken(token, getAllHeaders(request));
+          // Permissions on a type are the same as keyspace (CreateTypeFetcher for graphQl API)
+          db.getAuthorizationService()
+              .authorizeSchemaRead(
+                  authenticatedDB.getAuthenticationSubject(),
+                  Collections.singletonList(keyspaceName),
+                  null,
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:185]: When retrieving all types there's a collection here. Would be the case of setting there as null or using a singletonList here?

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+                  SourceAPI.REST,
+                  ResourceKind.TYPE);
+
+          authenticatedDB
+              .getDataStore()
+              .queryBuilder()
+              .drop()
+              .type(
+                  keyspaceName,
+                  ImmutableUserDefinedType.builder().keyspace(keyspaceName).name(typeName).build())
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:355]: Just double checking here. There's no way this can lead to NPE if the keyspace doesn't exists, right?

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java (hunk 5 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:347]: sure will update accordingly

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java (hunk 4 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:260]: ok

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java (hunk 3 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:270]: ok

File: restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+import io.swagger.annotations.ApiModelProperty;
+
+/**
+ * Represents a column in a User Defined type like {@link UserDefinedTypeAdd}
+ *
+ * <p>This class hase been introduced as the existing {@link ColumnDefinition} included attributes
+ * relative to Table only (static). Inheritance would still be possible.
+ *
+ * @since 1.0.16
+ * @author Cedrick LUNVEN (@clunven)
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java:30]: nit: I didn't find other places where there's a @author defined. Wouldn't be the case of removing it?

File: restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class ColumnDefinitionUserDefinedType {
+
+  /** (Immutable) unique identifier for the column. */
+  private final String name;
+
+  /** (Immutable) type definition like 'map<string,string>'. */
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java:38]: nit: Sort of redundant comment. Plus the example could be moved to the ApiModelProperty's value

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java (hunk 6 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:132]: Really good question. The RBAC in Cassandra is can be at Table/Type level so having this control would be relevant. 

Now `authorizeSchemaWrite()/authorizeSchemaRead()` should probably perform a filter and not a ok/ko returning only granted objects

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java (hunk 7 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:250]: True. On every single method to be honest and externalize as a private method IMO

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java (hunk 10 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:355]: Do not see any NPE in the `QueryBuilderImpl`, still checking existence before is required as stated above

File: restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  private final String typeDefinition;
```
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java:39]: nit: **maybe** renaming to just `type`, so the JSON can be `{"name": ... , "type": "set<int>"}`.

File: restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java (hunk 7 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/resources/v2/schemas/UserDefinedTypesResource.java:250]: There is a review needed in every resource most do not do the test.

File: restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java (hunk 12 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java:38]: If ApiModelProperty is introduced the backward compatibility is broker and V3 should be provided right ?

File: restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java (hunk 13 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java:39]: type is always tricky to marshall/unmarshall but could change it if needed

File: pom.xml (hunk 2 above)
Comment (mpenick) [reply pom.xml:367]: Not sure what you mean here.  Can you not generate that code with `mvn compile`? This is what I do when using intellij.  I would prefer not to have comment out code without explaining why it's there, but would prefer not to have it at all.

File: restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+ * Represents a column in a User Defined type like {@link UserDefinedTypeAdd}
+ *
+ * <p>This class hase been introduced as the existing {@link ColumnDefinition} included attributes
+ * relative to Table only (static). Inheritance would still be possible.
+ *
+ * @since 1.0.16
+ * @author Cedrick LUNVEN (@clunven)
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class ColumnDefinitionUserDefinedType {
```
Comment (mpenick) [inline restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java:33]: This seems to be used for defining UDT fields. Should it be called something like `FieldDefinitionUserDefinedType` or `UdtField`, etc.?

File: restapi/src/main/java/io/stargate/web/models/UserDefinedTypeAdd.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+  public void setIfNotExists(boolean ifNotExists) {
+    this.ifNotExists = ifNotExists;
+  }
+
+  /**
+   * Getter accessor for attribute 'columnDefinitions'.
+   *
+   * @return current value of 'columnDefinitions'
+   */
+  public List<ColumnDefinitionUserDefinedType> getColumnDefinitions() {
```
Comment (mpenick) [inline restapi/src/main/java/io/stargate/web/models/UserDefinedTypeAdd.java:64]: `getFieldDefinitions()` and `setFieldDefintions()`?

File: restapi/src/main/java/io/stargate/web/models/UserDefinedTypeResponse.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    return name;
+  }
+
+  @ApiModelProperty(value = "Name of the keyspace the user defined type belongs.")
+  public String getKeyspace() {
+    return keyspace;
+  }
+
+  @ApiModelProperty(value = "Definition of columns within the user defined type.")
+  public List<ColumnDefinitionUserDefinedType> getColumnDefinitions() {
```
Comment (mpenick) [inline restapi/src/main/java/io/stargate/web/models/UserDefinedTypeResponse.java:48]: "Column" --> "Field"

File: restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java (hunk 14 above)
Comment (clun) [reply restapi/src/main/java/io/stargate/web/models/ColumnDefinitionUserDefinedType.java:33]: ACK=> let's go with field, this week were packed full, will update the PR next tuesday

**Comment by eribeiro** [comment]
@clun Hi, I have closed my duplicate PR, but I think there are some snippets that could be ported to this PR (e.g., integration tests). Please, let me know if/where I can help. Cheers! :)

**Comment by clun** [comment]
@eribeiro I would grateful if you could take the lead here, unfortunately It seems I am a bottleneck, with a lot on my plate beside this great project.

**Comment by eribeiro** [comment]
@clun Sure mate, no problem. 🙂 I'll try to address the review comments in this PR and add some IT tests. Thanks!

**Comment by dougwettlaufer** [comment]
Work is moving to https://github.com/stargate/stargate/pull/895
